File: src/test/starterkit/starter.cs
#define SHOW_DEBUG

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace ACOIF
{
    public class ACodeOfIceAndFire
    {
        public enum BuildingType
        {
            Hq,
            Mine,
            Tower
        }

        public enum Team
        {
            Fire,
            Ice
        }

        private const int WIDTH = 12;
        private const int HEIGHT = 12;

        private const int ME = 0;
        private const int OPPONENT = 1;
        private const int NEUTRAL = -1;

        private const int TRAIN_COST_LEVEL_1 = 10;
        private const int TRAIN_COST_LEVEL_2 = 20;
        private const int TRAIN_COST_LEVEL_3 = 30;

        private static void Main()
        {
            var game = new Game();
            game.Init();

            // game loop
            while (true)
            {
                game.Update();
                game.Solve();
                Console.WriteLine(game.Output.ToString());
            }
        }

        public class Game
        {
            public readonly List<Building> Buildings = new List<Building>();

            public readonly Tile[,] Map = new Tile[WIDTH, HEIGHT];
            public readonly StringBuilder Output = new StringBuilder();

            // Not Usefull in Wood3
            public List<Position> MineSpots = new List<Position>();

            public int MyGold;
            public int MyIncome;
            public Team MyTeam;

            public int OpponentGold;
            public int OpponentIncome;
            public int Turn;
            public List<Unit> Units = new List<Unit>();

            public List<Unit> MyUnits => Units.Where(u => u.IsOwned).ToList();
            public List<Unit> OpponentUnits => Units.Where(u => u.IsOpponent).ToList();

            public Position MyHq => MyTeam == Team.Fire ? (0, 0) : (11, 11);
            public Position OpponentHq => MyTeam == Team.Fire ? (11, 11) : (0, 0);

            public List<Position> MyPositions = new List<Position>();
            public List<Position> OpponentPositions = new List<Position>();
            public List<Position> NeutralPositions = new List<Position>();

            public void Init()
            {
                for (var y = 0; y < HEIGHT; y++)
                for (var x = 0; x < WIDTH; x++)
                {
                    Map[x, y] = new Tile
                    {
                        Position = (x, y)
                    };
                }

                var numberMineSpots = int.Parse(Console.ReadLine());
                for (var i = 0; i < numberMineSpots; i++)
                {
                    var inputs = Console.ReadLine().Split(' ');
                    MineSpots.Add((int.Parse(inputs[0]), int.Parse(inputs[1])));
                }
            }

            public void Update()
            {
                Units.Clear();
                Buildings.Clear();

                MyPositions.Clear();
                OpponentPositions.Clear();
                NeutralPositions.Clear();

                Output.Clear();

                // --------------------------------------

                MyGold = int.Parse(Console.ReadLine());
                MyIncome = int.Parse(Console.ReadLine());
                OpponentGold = int.Parse(Console.ReadLine());
                OpponentIncome = int.Parse(Console.ReadLine());

                // Read Map
                for (var y = 0; y < HEIGHT; y++)
                {
                    var line = Console.ReadLine();
                    for (var x = 0; x < WIDTH; x++)
                    {
                        var c = line[x] + "";
                        Map[x, y].IsWall = c == "#";
                        Map[x, y].Active = "OX".Contains(c);
                        Map[x, y].Owner = c.ToLower() == "o" ? ME : c.ToLower() == "x" ? OPPONENT : NEUTRAL;
                        Map[x, y].HasMineSpot = MineSpots.Count(spot => spot == (x, y)) > 0;
                    
                        Position p = (x, y);
                        if(Map[x, y].IsOwned)
                            MyPositions.Add(p);
                        else if(Map[x, y].IsOpponent)
                            OpponentPositions.Add(p);
                        else if(!Map[x, y].IsWall) {
                            NeutralPositions.Add(p);
                        }
                    }
                }

                // Read Buildings
                var buildingCount = int.Parse(Console.ReadLine());
                for (var i = 0; i < buildingCount; i++)
                {
                    var inputs = Console.ReadLine().Split(' ');
                    Buildings.Add(new Building
                    {
                        Owner = int.Parse(inputs[0]),
                        Type = (BuildingType) int.Parse(inputs[1]),
                        Position = (int.Parse(inputs[2]), int.Parse(inputs[3]))
                    });
                }

                // Read Units
                var unitCount = int.Parse(Console.ReadLine());
                for (var i = 0; i < unitCount; i++)
                {
                    var inputs = Console.ReadLine().Split(' ');
                    Units.Add(new Unit
                    {
                        Owner = int.Parse(inputs[0]),
                        Id = int.Parse(inputs[1]),
                        Level = int.Parse(inputs[2]),
                        Position = (int.Parse(inputs[3]), int.Parse(inputs[4]))
                    });
                }

                // --------------------------------

                // Get Team
                MyTeam = Buildings.Find(b => b.IsHq && b.IsOwned).Position == (0, 0) ? Team.Fire : Team.Ice;

                // Usefull for symmetric AI
                if (MyTeam == Team.Ice) {
                    MyPositions.Reverse();
                    OpponentPositions.Reverse();
                    NeutralPositions.Reverse();
                }

                // --------------------------------

                // Debug
                Debug();
            }

            [Conditional("SHOW_DEBUG")]
            public void Debug()
            {
                Console.Error.WriteLine($"Turn: {Turn}");
                Console.Error.WriteLine($"My team: {MyTeam}");
                Console.Error.WriteLine($"My gold: {MyGold} (+{MyIncome})");
                Console.Error.WriteLine($"Opponent gold: {OpponentGold} (+{OpponentIncome})");

                Console.Error.WriteLine("=====");
                foreach (var b in Buildings) Console.Error.WriteLine(b);
                foreach (var u in Units) Console.Error.WriteLine(u);
            }

            /***
             * -----------------------------------------------------------
             * TODO Solve
             * -----------------------------------------------------------
             */
            public void Solve()
            {
                // Make sur the AI doesn't timeout
                Wait();

                MoveUnits();
                
                TrainUnits();
                
                Turn++;
            }
            
            public void MoveUnits()
            {
                // Rush center
                Position target = MyTeam == Team.Fire ? (5,5) : (6,6);

                if (Map[target.X, target.Y].IsOwned) return;

                foreach (var unit in MyUnits)
                    Move(unit.Id, target);
            }
            
            public void TrainUnits()
            {
                Position target = MyTeam == Team.Fire ? (1,0) : (10,11);

                if (MyGold >= TRAIN_COST_LEVEL_1)
                    Train(1, target);
            }

            public void Wait()
            {
                Output.Append("WAIT;");
            }

            public void Train(int level, Position position)
            {
                // TODO: Handle upkeep
                int cost = 0;
                switch (level) {
                    case 1 : cost = TRAIN_COST_LEVEL_1; break;
                    case 2 : cost = TRAIN_COST_LEVEL_2; break;
                    case 3 : cost = TRAIN_COST_LEVEL_3; break;
                }

                MyGold -= cost;
                Output.Append($"TRAIN {level} {position.X} {position.Y};");
            }

            public void Move(int id, Position position)
            {
                // TODO: Handle map change
                Output.Append($"MOVE {id} {position.X} {position.Y};");
            }
            
            // TODO: Handle Build command
        }


        public class Unit : Entity
        {
            public int Id;
            public int Level;

            public override string ToString() => $"Unit => {base.ToString()} Id: {Id} Level: {Level}";
        }

        public class Building : Entity
        {
            public BuildingType Type;

            public bool IsHq => Type == BuildingType.Hq;
            public bool IsTower => Type == BuildingType.Tower;
            public bool IsMine => Type == BuildingType.Mine;

            public override string ToString() => $"Building => {base.ToString()} Type: {Type}";
        }

        public class Entity
        {
            public int Owner;
            public Position Position;

            public bool IsOwned => Owner == ME;
            public bool IsOpponent => Owner == OPPONENT;

            public int X => Position.X;
            public int Y => Position.Y;

            public override string ToString() => $"Owner: {Owner} Position: {Position}";
        }

        public class Tile
        {
            public bool Active;
            public bool HasMineSpot;
            public bool IsWall;

            public int Owner = NEUTRAL;

            public Position Position;
            public int X => Position.X;
            public int Y => Position.Y;

            public bool IsOwned => Owner == ME;
            public bool IsOpponent => Owner == OPPONENT;
            public bool IsNeutral => Owner == NEUTRAL;
        }

        public class Position
        {
            public int X;
            public int Y;

            public static implicit operator Position(ValueTuple<int, int> cell) => new Position
            {
                X = cell.Item1,
                Y = cell.Item2
            };

            public override string ToString() => $"({X},{Y})";

            public static bool operator ==(Position obj1, Position obj2) => obj1.Equals(obj2);

            public static bool operator !=(Position obj1, Position obj2) => !obj1.Equals(obj2);

            public override bool Equals(object obj) => Equals((Position) obj);

            protected bool Equals(Position other) => X == other.X && Y == other.Y;

            public double Dist(Position p) => Math.Abs(X - p.X) + Math.Abs(Y - p.Y);
        }
    }
}

File: src/test/starterkit/starter.vb
#Const SHOW_DEBUG = True

Imports System.Collections.Generic
Imports System.Diagnostics

Module ACOIF

    Public Enum BuildingType
        Hq
        Mine
        Tower
    End Enum

    Public Enum Team
        Fire
        Ice
    End Enum

    Private Const WIDTH As Integer = 12
    Private Const HEIGHT As Integer = 12

    Private Const MYSELF As Integer = 0
    Private Const OPPONENT As Integer = 1
    Private Const NEUTRAL As Integer = -1

    Private Const TRAIN_COST_LEVEL_1 As Integer = 10
    Private Const TRAIN_COST_LEVEL_2 As Integer = 20
    Private Const TRAIN_COST_LEVEL_3 As Integer = 30

    Sub Main()

        Dim game As New Game()
        game.Init()

        ' game loop
        While True
            game.Update()
            game.Solve()
            Console.WriteLine(game.Output.ToString())
        End While
    End Sub

    Public Class Game
        Public ReadOnly Buildings As New List(Of Building)

        Public ReadOnly Map(WIDTH, HEIGHT) As Tile
        Public Output As String = ""

        ' Not Usefull in Wood3
        Public MineSpots As New List(Of Position)

        Public MyGold As Integer
        Public MyIncome As Integer
        Public MyTeam As Team

        Public OpponentGold As Integer
        Public OpponentIncome As Integer
        Public Turn As Integer
        Public Units As New List(Of Unit)

        Public Function MyUnits() As List(Of Unit)
            Dim r As New List(Of Unit)
            For Each u As Unit In Units
                If u.IsOwned Then r.Add(u)
            Next
            Return r
        End Function
        Public Function OpponentUnits() As List(Of Unit)
            Dim r As New List(Of Unit)
            For Each u As Unit In Units
                If u.IsOpponent Then r.Add(u)
            Next
            Return r
        End Function

        Public Function MyHq() As Position
            If MyTeam = Team.Fire Then
                Return New Position(0, 0)
            Else
                Return New Position(11, 11)
            End If
        End Function

        Public Function OpponentHq() As Position
            If MyTeam = Team.Fire Then
                Return New Position(11, 11)
            Else
                Return New Position(0, 0)
            End If
        End Function

        Public MyPositions As New List(Of Position)
        Public OpponentPositions As New List(Of Position)
        Public NeutralPositions As New List(Of Position)

        Public Sub Init()
            For y As Integer = 0 To HEIGHT - 1
                For x As Integer = 0 To WIDTH - 1
                    Dim t As New Tile
                    t.Position = New Position(x, y)
                    Map(x, y) = t
                Next
            Next

            Dim numberMineSpots As Integer = CInt(Console.ReadLine())
            For i As Integer = 0 To numberMineSpots - 1
                Dim inputs() As String = Console.ReadLine().Split(" "c)
                MineSpots.Add(New Position(CInt(inputs(0)), CInt(inputs(1))))
            Next
        End Sub

        Public Sub Update()
            Units.Clear()
            Buildings.Clear()

            MyPositions.Clear()
            OpponentPositions.Clear()
            NeutralPositions.Clear()

            Output = ""

            ' --------------------------------------

            MyGold = CInt(Console.ReadLine())
            MyIncome = CInt(Console.ReadLine())
            OpponentGold = CInt(Console.ReadLine())
            OpponentIncome = CInt(Console.ReadLine())

            ' Read Map
            For y As Integer = 0 To HEIGHT - 1
                Dim line As String = Console.ReadLine()
                For x As Integer = 0 To WIDTH - 1
                    Dim c As String = line(x) & ""
                    Map(x, y).IsWall = c = "#"
                    Map(x, y).Active = "OX".Contains(c)
                    Map(x, y).Owner = If(c.ToLower() = "o", MYSELF, If(c.ToLower() = "x", OPPONENT, NEUTRAL))
                    Map(x, y).HasMineSpot = MineSpots.Contains(New Position(x, y))

                    Dim p As New Position(x, y)
                    If Map(x, y).IsOwned Then
                        MyPositions.Add(p)
                    ElseIf Map(x, y).IsOpponent Then
                        OpponentPositions.Add(p)
                    ElseIf Not Map(x, y).IsWall Then
                        NeutralPositions.Add(p)
                    End If
                Next
            Next

            ' Read Buildings
            Dim buildingCount As Integer = CInt(Console.ReadLine())
            For i As Integer = 0 To buildingCount - 1
                Dim inputs() As String = Console.ReadLine().Split(" "c)
                Dim b As New Building
                b.Owner = CInt(inputs(0))
                b.Type = DirectCast(CInt(inputs(1)), BuildingType)
                b.Position = New Position(CInt(inputs(2)), CInt(inputs(3)))
                Buildings.Add(b)
            Next

            ' Read Units
            Dim unitCount As Integer = CInt(Console.ReadLine())
            For i As Integer = 0 To unitCount - 1
                Dim inputs() As String = Console.ReadLine().Split(" "c)
                Dim u As New Unit
                u.Owner = CInt(inputs(0))
                u.Id = CInt(inputs(1))
                u.Level = CInt(inputs(2))
                u.Position = New Position(CInt(inputs(3)), CInt(inputs(4)))
                Units.Add(u)
            Next

            ' --------------------------------

            ' Get Team
            MyTeam = Team.Ice
            For Each b As Building In Buildings
                If b.IsOwned AndAlso b.IsHq AndAlso b.Position = New Position(0, 0) Then
                    MyTeam = Team.Fire : Exit For
                End If
            Next

            ' Usefull for symmetric AI
            If MyTeam = Team.Ice Then
                MyPositions.Reverse()
                OpponentPositions.Reverse()
                NeutralPositions.Reverse()
            End If

            ' --------------------------------

            ' Debug
            Debug()
        End Sub

        <Conditional("SHOW_DEBUG")>
        Public Sub Debug()
            Console.Error.WriteLine("Turn: " & Turn)
            Console.Error.WriteLine("My team: " & MyTeam)
            Console.Error.WriteLine("My gold: " & MyGold & " (+" & MyIncome & ")")
            Console.Error.WriteLine("Opponent gold: " & OpponentGold & " (+" & OpponentIncome & ")")

            Console.Error.WriteLine("=====")
            For Each b As Building In Buildings
                Console.Error.WriteLine(b)
            Next

            For Each u As Unit In Units
                Console.Error.WriteLine(u)
            Next
        End Sub

        '-----------------------------------------------------------
        'TODO Solve
        '-----------------------------------------------------------

        Public Sub Solve()
            ' Make sur the AI doesn't timeout
            Wait()

            MoveUnits()

            TrainUnits()

            Turn += 1
        End Sub

        Public Sub MoveUnits()
            ' Rush center
            Dim target As Position
            If MyTeam = Team.Fire Then
                target = New Position(5, 5)
            Else
                target = New Position(6, 6)
            End If

            If (Map(target.X, target.Y).IsOwned) Then Exit Sub

            For Each u As Unit In MyUnits()
                Move(u.Id, target)
            Next
        End Sub

        Public Sub TrainUnits()
            Dim target As Position
            If MyTeam = Team.Fire Then
                target = New Position(1, 0)
            Else
                target = New Position(10, 11)
            End If

            If MyGold >= TRAIN_COST_LEVEL_1 Then
                Train(1, target)
            End If
        End Sub

        Public Sub Wait()
            Output &= "WAIT;"
        End Sub

        Public Sub Train(level As Integer, position As Position)
            ' TODO: Handle upkeep
            Dim cost As Integer = 0
            Select Case level
                Case 1 : cost = TRAIN_COST_LEVEL_1
                Case 2 : cost = TRAIN_COST_LEVEL_2
                Case 3 : cost = TRAIN_COST_LEVEL_3
            End Select

            MyGold -= cost
            Output &= "TRAIN " & level & " " & position.X & " " & position.Y & ";"
        End Sub

        Public Sub Move(id As Integer, position As Position)
            ' TODO: Handle map change
            Output &= "MOVE " & id & " " & position.X & " " & position.Y & ";"
        End Sub

        ' TODO: Handle Build command
    End Class


    Public Class Unit
        Inherits Entity
        Public Id As Integer
        Public Level As Integer

        Public Overrides Function toString() As String
            Return "Unit => " & MyBase.toString() & " Id: " & Id & " Level: " & Level
        End Function
    End Class

    Public Class Building
        Inherits Entity

        Public Type As BuildingType

        ReadOnly Property IsHq As Boolean
            Get
                Return Type = BuildingType.Hq
            End Get
        End Property

        ReadOnly Property IsTower As Boolean
            Get
                Return Type = BuildingType.Tower
            End Get
        End Property

        ReadOnly Property IsMine As Boolean
            Get
                Return Type = BuildingType.Mine
            End Get
        End Property

        Public Overrides Function toString() As String
            Return "Building => " & MyBase.toString() & " Type: " & Type
        End Function
    End Class

    Public Class Entity
        Public Owner As Integer
        Public Position As Position

        ReadOnly Property IsOwned As Boolean
            Get
                Return Owner = MYSELF
            End Get
        End Property

        ReadOnly Property IsOpponent As Boolean
            Get
                Return Owner = OPPONENT
            End Get
        End Property

        ReadOnly Property X As Integer
            Get
                Return Position.X
            End Get
        End Property

        ReadOnly Property Y As Integer
            Get
                Return Position.Y
            End Get
        End Property

        Public Overrides Function toString() As String
            Return "Owner: " & Owner & " Position: " & Position.toString
        End Function

    End Class

    Public Class Tile
        Public Active As Boolean
        Public HasMineSpot As Boolean
        Public IsWall As Boolean

        Public Owner As Integer = NEUTRAL

        Public Position As Position
        ReadOnly Property X As Integer
            Get
                Return Position.X
            End Get
        End Property

        ReadOnly Property Y As Integer
            Get
                Return Position.Y
            End Get
        End Property

        Public Function IsOwned() As Boolean
            Return Owner = MYSELF
        End Function
        Public Function IsOpponent() As Boolean
            Return Owner = OPPONENT
        End Function
        Public Function IsNeutral() As Boolean
            Return Owner = NEUTRAL
        End Function
    End Class

    Public Structure Position
        Implements IEquatable(Of Position)

        Public X As Integer
        Public Y As Integer

        Sub New(X As Integer, Y As Integer)
            Me.X = X
            Me.Y = Y
        End Sub

        Public Overrides Function toString() As String
            Return "(" & X & "," & Y & ")"
        End Function

        Public Shared Operator =(obj1 As Position, obj2 As Position) As Boolean
            Return obj1.Equals(obj2)
        End Operator

        Public Shared Operator <>(obj1 As Position, obj2 As Position) As Boolean
            Return Not obj1.Equals(obj2)
        End Operator

        Public Overrides Function Equals(obj As Object) As Boolean
            If TypeOf obj Is Position Then
                Return Equals(DirectCast(obj, Position))
            Else
                Return False
            End If
        End Function

        Public Overloads Function Equals(other As Position) As Boolean Implements IEquatable(Of Position).Equals
            Return X = other.X AndAlso Y = other.Y
        End Function

        Public Function Dist(p As Position) As Double
            Return Math.Abs(X - p.X) + Math.Abs(Y - p.Y)
        End Function
    End Structure

End Module